src/include/ceph_features.h

```diff
@@ -65,6 +65,21 @@
  *
  * This ensures that no two versions who have different meanings for
  * the bit ever speak to each other.
+ *
+ * 2) For feature bits that were *anyhow* exposed to clients (even through
+ * a common part like OSDMap), the requirements are stricter: instead of
+ * the n-2 rule we have for the inter-daemon communication, for clients
+ * we *guarantee* compatibility up n-3 major upstream versions. To exemplify:
```

Comment:
Hmm, if the DEPRECATED macros and MASK stuff doesn't make it "do the right thing" for random clients, I think feature bits which are client-visible are just...not reusable or able to be retired.

@athanatos @jdurgin do either of you remember much of the planning around feature bit retirement? I don't think I got involved in its development, but I think I would have noticed if we went from "all clients meeting the admin-configurable min_compat_client" to "3 versions, and incredibly messy for the kernel". I'm sure @idryomov has thoughts about that.

Reply:
Hrm, I don't remember, I'd need to take a bit of time and think it through.

Reply:
Client-facing feature bits can't really be retired in an orderly fashion, so any `N-<something>` rule is bound to not work, exactly because

> we have little control over kernel versions and user-space libraries within third party's containers (old librbd, ceph-fuse and so on).

We have done this only once before, slurping up all argonaut-ish feature bits.  I suppose we could do this again for pre-hammer feature bits some time soon but this would have to be an ad-hoc, review each feature bit sort of thing.  If the intent is to document the procedure, I would rather document that there is no procedure and go with something like

> For feature bits that were *anyhow* exposed to clients (even through a common part like OSDMap): don't touch the ones that are already there, think hard before introducing a new one, assume a lifetime of at least 10 years.

Reply:
Thanks for the input, Ilya!

My intention is just to document the state of the thing – discover & document. I'm aware that another views (i.e. _n - 3_ as as **hard** guarantee) also exist. I think it would be good to describe also how `mon_osd_initial_require_min_compat_client` interplays here.

CC: @gregsfortytwo, @athanatos, @jdurgin.

Reply:
`require_min_compat_client` is only tangentially related: it's a setting that is intended to prevent an administrator from accidentally enabling features that older clients would choke on (i.e. features that would prevent older clients from connection to the cluster).  The fact that it's set to luminous by default doesn't mean much: in the end it's based on the same feature bits, many of which can be manipulated in isolation.